modin/backends/base/doc_utils.py

```diff
@@ -0,0 +1,685 @@
+
+from functools import partial
+from modin.utils import append_to_docstring, format_string
+
+
+_one_column_warning = """
+.. warning::
+    This method is supported only by one-column query compilers.
+"""
+
+_deprecation_warning = """
+.. warning::
+    This method duplicates logic of ``{0}`` and will be removed soon.
+"""
+
+_refer_to_note = """
+Notes
+-----
+Please refer to ``modin.pandas.{0}`` for more information
+about parameters and output format.
+"""
+
+add_one_column_warning = append_to_docstring(_one_column_warning)
+
+
+def add_deprecation_warning(replacement_method):
+    """
+    Build decorator which appends deprecation warning to the function's docstring.
+
+    Appended warning indicates that the current method duplicates functionality of
+    some other method and so is slated to be removed in the future.
+
+    Parameters
+    ----------
+    replacement_method : str
+        Name of the method to use instead of deprecated.
+
+    Returns
+    -------
+    callable
+    """
+    message = _deprecation_warning.format(replacement_method)
+    return append_to_docstring(message)
+
+
+def add_refer_to(method):
+    """
+    Build decorator which appends link to the high-level equivalent method to the function's docstring.
+
+    Parameters
+    ----------
+    method : str
+        Method name in ``modin.pandas`` module to refer to.
+
+    Returns
+    -------
+    callable
+    """
+    # FIXME: this would break numpydoc if there already is a `Notes` section
+    note = _refer_to_note.format(method)
+    return append_to_docstring(note)
+
+
+def doc_qc_method(
+    template,
+    params=None,
+    refer_to=None,
+    refer_to_module_name=None,
+    one_column_method=False,
+    **kwargs,
+):
+    """
+    Build decorator which adds docstring for query compiler method.
+
+    Parameters
+    ----------
+    template : str
+        Method docstring in the NumPy docstyle format. Must contains {params}
+        placeholder.
+    params : str, optional
+        Method parameters in the NumPy docstyle format to substitute
+        in the `template`. `params` string should not include the "Parameters"
+        header.
+    refer_to : str, optional
+        Method name in `refer_to_module_name` module to refer to for more information
+        about parameters and output format.
+    refer_to_module_name : str, optional
+    one_column_method : bool, default: False
+        Whether to append note that this method is for one-column
+        query compilers only.
+    **kwargs : dict
+        Values to substitute in the `template`.
+
+    Returns
+    -------
+    callable
+    """
+    params_template = """
+
+        Parameters
+        ----------
+        {params}
+        """
+
+    params = format_string(params_template, params=params) if params else ""
+    substituted = format_string(template, params=params, refer_to=refer_to, **kwargs)
+    if refer_to_module_name:
+        refer_to = f"{refer_to_module_name}.{refer_to}"
```

Comment:
An entity name always like "module_name.entity", isn't it?

Reply:
but you're also using `refer_to` in its short form in line 120, why?

Reply:
In line 120 we're substituting raw values to the template, they shouldn't be processed anyhow, because when the source template contains `{refer_to}` placeholder we expect that the actual `refer_to` parameter value will be substituted. The actual refer-to note is appended [further](https://github.com/dchigarev/modin/blob/016e44f8d54354bdfc75ef8812732807b60ca51e/modin/backends/base/doc_utils.py#L127-L128) with the proper code entity name.
```python
from modin.backends.base.doc_utils import doc_qc_method

def foo():
    pass

print(
    doc_qc_method(
        template="Perform {refer_to} operation.",
        refer_to="foo",
        refer_to_module_name="bar",
    )(foo).__doc__
)

# Output:
# Perform foo operation.
#
# Notes
# -----
# Please refer to ``modin.pandas.bar.foo`` for more information
# about parameters and output format.
```

Reply:
So this looks like somewhat bad naming, where you're using `refer_to` both for referring to something else and to name the exact operation... however this is really minor, so we can leave it be for now.